File: csharp/TcpChannel.cs
using System.Net;
using System.Net.Sockets;

namespace Borkfork.SpiceDb.Embedded;

/// <summary>
///     Builds a gRPC SocketsHttpHandler that connects via TCP (host:port).
/// </summary>
internal static class TcpChannel
{
    public static SocketsHttpHandler CreateHandler(string address)
    {
        return new SocketsHttpHandler
        {
            ConnectCallback = async (_, ct) =>
            {
                string host;
                int port;
                if (address.StartsWith("["))
                {
                    // IPv6: [host]:port
                    var closeBracket = address.IndexOf(']');
                    if (closeBracket < 0 || closeBracket + 1 >= address.Length || address[closeBracket + 1] != ':')
                        throw new SpiceDbException("Invalid TCP address: " + address);
                    host = address.Substring(1, closeBracket - 1);
                    if (string.IsNullOrEmpty(host))
                        throw new SpiceDbException("Invalid TCP address (empty host): " + address);
                    if (!int.TryParse(address.Substring(closeBracket + 2), out port))
                        throw new SpiceDbException("Invalid TCP address (bad port): " + address);
                }
                else
                {
                    // IPv4 or hostname: host:port
                    var colonIdx = address.LastIndexOf(':');
                    if (colonIdx <= 0 || !int.TryParse(address.Substring(colonIdx + 1), out port))
                        throw new SpiceDbException("Invalid TCP address: " + address);
                    host = address.Substring(0, colonIdx);
                }

                var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
                await socket.ConnectAsync(new DnsEndPoint(host, port), ct);
                return new NetworkStream(socket, true);
            }
        };
    }
}

File: java/src/main/java/com/borkfork/spicedb/embedded/TcpChannel.java
package com.borkfork.spicedb.embedded;

import io.grpc.ManagedChannel;
import io.grpc.netty.NettyChannelBuilder;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.nio.NioSocketChannel;
import java.net.InetSocketAddress;

/** Builds a gRPC ManagedChannel that connects via TCP (host:port). */
final class TcpChannel {

  private TcpChannel() {}

  static ManagedChannel build(String address) {
    String host;
    int port;
    if (address.startsWith("[")) {
      // IPv6: [host]:port
      int closeBracket = address.indexOf(']');
      if (closeBracket < 0) {
        throw new IllegalArgumentException("Invalid TCP address (missing ']'): " + address);
      }
      host = address.substring(1, closeBracket);
      if (host.isEmpty()) {
        throw new IllegalArgumentException("Invalid TCP address (empty host): " + address);
      }
      if (closeBracket + 1 >= address.length() || address.charAt(closeBracket + 1) != ':') {
        throw new IllegalArgumentException("Invalid TCP address (missing port): " + address);
      }
      try {
        port = Integer.parseInt(address.substring(closeBracket + 2));
      } catch (NumberFormatException e) {
        throw new IllegalArgumentException("Invalid TCP address (bad port): " + address, e);
      }
    } else {
      // IPv4 or hostname: host:port
      int colon = address.lastIndexOf(':');
      if (colon <= 0) {
        throw new IllegalArgumentException("Invalid TCP address: " + address);
      }
      host = address.substring(0, colon);
      try {
        port = Integer.parseInt(address.substring(colon + 1));
      } catch (NumberFormatException e) {
        throw new IllegalArgumentException("Invalid TCP address (bad port): " + address, e);
      }
    }
    return NettyChannelBuilder.forAddress(new InetSocketAddress(host, port))
        .eventLoopGroup(new NioEventLoopGroup())
        .channelType(NioSocketChannel.class)
        .usePlaintext()
        .build();
  }
}
